File: runtime-go/retry.go
package runtime

import (
	"context"
	"errors"
	"math/rand"
	"sync/atomic"
	"time"
)

// ---------------------------------------------------------------------------
// Circuit Breaker
// ---------------------------------------------------------------------------

// CircuitState represents the three states of a circuit breaker.
type CircuitState int32

const (
	CircuitClosed   CircuitState = iota // Normal: requests pass through
	CircuitOpen                         // Tripped: requests fast-fail immediately
	CircuitHalfOpen                     // Probe: one request is allowed through to test recovery
)

func (cs CircuitState) String() string {
	switch cs {
	case CircuitClosed:
		return "CLOSED"
	case CircuitOpen:
		return "OPEN"
	case CircuitHalfOpen:
		return "HALF-OPEN"
	default:
		return "UNKNOWN"
	}
}

// CircuitBreaker is a thread-safe, atomic state machine that prevents cascading
// failures by fast-failing requests when the error rate exceeds a threshold.
type CircuitBreaker struct {
	state        int32  // CircuitState stored atomically
	failures     int64  // rolling failure count (atomic)
	successes    int64  // rolling success count in HalfOpen (atomic)
	lastOpenTime int64  // UnixNano of when circuit opened (atomic)

	// Config
	MaxFailures      int64         // failures before tripping (e.g. 5)
	OpenTimeout      time.Duration // how long to stay Open before probing
	HalfOpenProbes   int64         // successes required in HalfOpen to re-close
}

// NewCircuitBreaker returns a production-ready circuit breaker.
func NewCircuitBreaker(maxFailures int64, openTimeout time.Duration, halfOpenProbes int64) *CircuitBreaker {
	return &CircuitBreaker{
		MaxFailures:    maxFailures,
		OpenTimeout:    openTimeout,
		HalfOpenProbes: halfOpenProbes,
	}
}

// State returns the current CircuitState.
func (cb *CircuitBreaker) State() CircuitState {
	return CircuitState(atomic.LoadInt32(&cb.state))
}

// Allow returns true if the request should be allowed through.
func (cb *CircuitBreaker) Allow() bool {
	switch cb.State() {
	case CircuitClosed:
		return true
	case CircuitOpen:
		// After OpenTimeout, transition to HalfOpen for a probe attempt.
		openedAt := time.Unix(0, atomic.LoadInt64(&cb.lastOpenTime))
		if time.Since(openedAt) >= cb.OpenTimeout {
			if atomic.CompareAndSwapInt32(&cb.state, int32(CircuitOpen), int32(CircuitHalfOpen)) {
				atomic.StoreInt64(&cb.successes, 0)
			}
			return true
		}
		return false
	case CircuitHalfOpen:
		// Allow only one concurrent probe
		return true
	}
	return false
}

// RecordSuccess records a successful operation, potentially closing a HalfOpen circuit.
func (cb *CircuitBreaker) RecordSuccess() {
	switch cb.State() {
	case CircuitHalfOpen:
		probes := atomic.AddInt64(&cb.successes, 1)
		if probes >= cb.HalfOpenProbes {
			atomic.StoreInt32(&cb.state, int32(CircuitClosed))
			atomic.StoreInt64(&cb.failures, 0)
		}
	case CircuitClosed:
		// Reset failure count on success to prevent slow-burn tripping.
		atomic.StoreInt64(&cb.failures, 0)
	}
}

// RecordFailure records a failed operation, potentially opening the circuit.
func (cb *CircuitBreaker) RecordFailure() {
	f := atomic.AddInt64(&cb.failures, 1)
	if f >= cb.MaxFailures {
		if atomic.CompareAndSwapInt32(&cb.state, int32(CircuitClosed), int32(CircuitOpen)) ||
			atomic.CompareAndSwapInt32(&cb.state, int32(CircuitHalfOpen), int32(CircuitOpen)) {
			atomic.StoreInt64(&cb.lastOpenTime, time.Now().UnixNano())
		}
	}
}

// ---------------------------------------------------------------------------
// Token Bucket (for hedging rate limiting)
// ---------------------------------------------------------------------------

// TokenBucket implements a thread-safe token bucket for rate-limiting hedged
// requests. Tokens refill continuously at a fixed rate using atomic CAS.
type TokenBucket struct {
	// tokens is stored as a fixed-point integer: actual tokens = tokens / 1e9
	tokens    int64 // scaled by 1e9 (nano-tokens)
	capacity  int64 // max nano-tokens
	refillRate int64 // nano-tokens added per nanosecond
	lastRefill int64 // UnixNano of last refill (atomic)
}

// NewTokenBucket creates a bucket with `capacity` max tokens refilling at
// `ratePerSecond` tokens/second.
func NewTokenBucket(capacity float64, ratePerSecond float64) *TokenBucket {
	capNano := int64(capacity * 1_000_000_000)
	// refillRate: nano-tokens added per nanosecond = ratePerSecond / 1e9 * 1e9 = ratePerSecond
	refillRate := int64(ratePerSecond)
	if refillRate == 0 && ratePerSecond > 0 {
		refillRate = 1
	}
	return &TokenBucket{
		tokens:     capNano,
		capacity:   capNano,
		refillRate: refillRate,
		lastRefill: time.Now().UnixNano(),
	}
}



// NewHedgingTokenBucket returns a bucket that allows `hedgesPerSecond` hedged
// requests per second with a burst capacity of `burstCapacity`.
func NewHedgingTokenBucket(burstCapacity int, hedgesPerSecond float64) *TokenBucket {
	capNano := int64(burstCapacity) * 1_000_000_000
	// refillRate: nano-tokens added per nanosecond
	refillRate := int64(hedgesPerSecond * 1_000_000_000 / 1_000_000_000)
	if refillRate == 0 && hedgesPerSecond > 0 {
		refillRate = 1
	}
	return &TokenBucket{
		tokens:     capNano,
		capacity:   capNano,
		refillRate: refillRate,
		lastRefill: time.Now().UnixNano(),
	}
}

// Consume attempts to take one token. Returns true if successful.
func (tb *TokenBucket) Consume() bool {
	now := time.Now().UnixNano()
	last := atomic.LoadInt64(&tb.lastRefill)
	elapsed := now - last
	if elapsed > 0 {
		refill := elapsed * tb.refillRate
		if refill > 0 {
			for {
				cur := atomic.LoadInt64(&tb.tokens)
				next := cur + refill
				if next > tb.capacity {
					next = tb.capacity
				}
				if atomic.CompareAndSwapInt64(&tb.tokens, cur, next) {
					atomic.CompareAndSwapInt64(&tb.lastRefill, last, now)
					break
				}
			}
		}
	}
	// Attempt to consume one full token (1e9 nano-tokens)
	const oneToken = 1_000_000_000
	for {
		cur := atomic.LoadInt64(&tb.tokens)
		if cur < oneToken {
			return false
		}
		if atomic.CompareAndSwapInt64(&tb.tokens, cur, cur-oneToken) {
			return true
		}
	}
}

// ---------------------------------------------------------------------------
// RetryPolicy and ExecuteWithRetry
// ---------------------------------------------------------------------------

// RetryPolicy defines how failures are retried and how hedging is controlled.
type RetryPolicy struct {
	MaxAttempts       int
	InitialBackoff    time.Duration
	MaxBackoff        time.Duration
	BackoffMultiplier float64
	// RetryableErrors decides if an error warrants a retry. nil = always retry.
	RetryableErrors func(error) bool

	// HedgeDelay > 0 enables P99 hedging. A duplicate request fires if the
	// first does not complete within this window.
	HedgeDelay time.Duration

	// HedgeBucket limits the rate at which hedged requests may fire to prevent
	// thundering-herd amplification. nil = unlimited hedging (dangerous).
	HedgeBucket *TokenBucket

	// Breaker, if non-nil, fast-fails requests when the circuit is open.
	Breaker *CircuitBreaker
}

// DefaultRetryPolicy returns a production-safe policy with exponential backoff,
// a 10 hedges/s token bucket, and a 5-failure circuit breaker.
func DefaultRetryPolicy() RetryPolicy {
	return RetryPolicy{
		MaxAttempts:       3,
		InitialBackoff:    100 * time.Millisecond,
		MaxBackoff:        2 * time.Second,
		BackoffMultiplier: 2.0,
		RetryableErrors:   nil, // retry all errors
		HedgeDelay:        0,
		HedgeBucket:       NewHedgingTokenBucket(5, 10),
		Breaker:           NewCircuitBreaker(5, 30*time.Second, 2),
	}
}

// ExecuteWithRetry runs the operation with the given policy.
func ExecuteWithRetry[T any](ctx context.Context, policy RetryPolicy, operation func(context.Context) (T, error)) (T, error) {
	// Circuit breaker fast-fail
	if policy.Breaker != nil && !policy.Breaker.Allow() {
		var zero T
		return zero, errors.New("circuit open: too many recent failures")
	}

	if policy.HedgeDelay > 0 {
		return executeWithHedging(ctx, policy, operation)
	}
	return executeRetriesOnly(ctx, policy, operation)
}

func executeRetriesOnly[T any](ctx context.Context, policy RetryPolicy, operation func(context.Context) (T, error)) (T, error) {
	var lastErr error
	backoff := policy.InitialBackoff

	for attempt := 1; attempt <= policy.MaxAttempts; attempt++ {
		result, err := operation(ctx)
		if err == nil {
			if policy.Breaker != nil {
				policy.Breaker.RecordSuccess()
			}
			return result, nil
		}

		if policy.Breaker != nil {
			policy.Breaker.RecordFailure()
		}
		if policy.RetryableErrors != nil && !policy.RetryableErrors(err) {
			return result, err
		}
		lastErr = err

		if attempt == policy.MaxAttempts {
			break
		}

		// Exponential backoff with full-jitter (avoids thundering herd on retry)
		jitterRange := int64(backoff)
		jitter := time.Duration(rand.Int63n(jitterRange))
		sleepDuration := jitter // full-jitter: sleep = rand[0, backoff)

		select {
		case <-time.After(sleepDuration):
		case <-ctx.Done():
			var zero T
			return zero, ctx.Err()
		}

		nextBackoff := time.Duration(float64(backoff) * policy.BackoffMultiplier)
		if nextBackoff > policy.MaxBackoff {
			nextBackoff = policy.MaxBackoff
		}
		backoff = nextBackoff
	}

	var zero T
	return zero, lastErr
}

// executeWithHedging fires a duplicate request if the primary exceeds HedgeDelay.
// The token bucket prevents hedging from amplifying load during a cluster-wide
// latency spike.
func executeWithHedging[T any](ctx context.Context, policy RetryPolicy, operation func(context.Context) (T, error)) (T, error) {
	type result struct {
		val T
		err error
	}

	resultCh := make(chan result, 2)

	// Primary attempt
	primaryCtx, primaryCancel := context.WithCancel(ctx)
	defer primaryCancel()

	go func() {
		v, e := executeRetriesOnly(primaryCtx, policy, operation)
		resultCh <- result{v, e}
	}()

	select {
	case res := <-resultCh:
		return res.val, res.err

	case <-time.After(policy.HedgeDelay):
		// Only hedge if the token bucket has capacity.
		canHedge := policy.HedgeBucket == nil || policy.HedgeBucket.Consume()
		if !canHedge {
			// Token exhausted: wait for primary to finish.
			select {
			case res := <-resultCh:
				return res.val, res.err
			case <-ctx.Done():
				var zero T
				return zero, ctx.Err()
			}
		}

		// Launch hedged request on a fresh context so it can outlive the primary.
		hedgeCtx, hedgeCancel := context.WithCancel(ctx)
		defer hedgeCancel()

		go func() {
			v, e := executeRetriesOnly(hedgeCtx, policy, operation)
			resultCh <- result{v, e}
		}()

		// Accept whichever succeeds first; cancel the loser.
		res1 := <-resultCh
		if res1.err == nil {
			return res1.val, nil
		}
		res2 := <-resultCh
		if res2.err == nil {
			return res2.val, nil
		}
		var zero T
		return zero, errors.Join(res1.err, res2.err)

	case <-ctx.Done():
		var zero T
		return zero, ctx.Err()
	}
}

File: runtimes/python/helix_rt/retry.py
"""
helix_rt.retry
~~~~~~~~~~~~~~
Exponential backoff, Circuit Breaker, Token Bucket, and P99 Hedging —
full Python port of the Go runtime-go/retry.go implementation.

Usage::

    from helix_rt.retry import RetryPolicy, execute_with_retry, CircuitBreaker

    policy = RetryPolicy.with_defaults()

    async def call_model():
        ...

    result = await execute_with_retry(policy, call_model)
"""

from __future__ import annotations

import asyncio
import random
import time
from dataclasses import dataclass, field
from enum import IntEnum
from typing import Callable, Awaitable, Optional, TypeVar

T = TypeVar("T")


# ---------------------------------------------------------------------------
# Circuit Breaker
# ---------------------------------------------------------------------------

class CircuitState(IntEnum):
    CLOSED    = 0
    OPEN      = 1
    HALF_OPEN = 2


class CircuitBreaker:
    """
    Thread-safe (asyncio-safe) circuit breaker matching the Go/Rust implementations.

    State machine:
      CLOSED  → normal traffic
      OPEN    → fast-fail; re-checks after open_timeout
      HALF_OPEN → one probe request; closes on success, reopens on failure
    """

    def __init__(
        self,
        max_failures: int = 5,
        open_timeout: float = 30.0,  # seconds
        half_open_probes: int = 2,
    ) -> None:
        self.max_failures     = max_failures
        self.open_timeout     = open_timeout
        self.half_open_probes = half_open_probes

        self._state     = CircuitState.CLOSED
        self._failures  = 0
        self._successes = 0
        self._opened_at = 0.0
        self._lock      = asyncio.Lock()

    async def state(self) -> CircuitState:
        async with self._lock:
            return self._state

    async def allow(self) -> bool:
        async with self._lock:
            if self._state == CircuitState.CLOSED:
                return True
            if self._state == CircuitState.HALF_OPEN:
                return True
            # OPEN: check timeout
            if time.monotonic() - self._opened_at >= self.open_timeout:
                self._state     = CircuitState.HALF_OPEN
                self._successes = 0
                return True
            return False

    async def record_success(self) -> None:
        async with self._lock:
            if self._state == CircuitState.HALF_OPEN:
                self._successes += 1
                if self._successes >= self.half_open_probes:
                    self._state    = CircuitState.CLOSED
                    self._failures = 0
            elif self._state == CircuitState.CLOSED:
                self._failures = 0

    async def record_failure(self) -> None:
        async with self._lock:
            self._failures += 1
            if self._failures >= self.max_failures and self._state == CircuitState.CLOSED:
                self._state     = CircuitState.OPEN
                self._opened_at = time.monotonic()


# ---------------------------------------------------------------------------
# Token Bucket (hedging rate limiter)
# ---------------------------------------------------------------------------

class TokenBucket:
    """
    Thread-safe token bucket for limiting hedged request rate.
    Prevents hedge-induced thundering-herd during cluster-wide latency spikes.
    """

    def __init__(self, capacity: int, rate_per_second: float) -> None:
        self.capacity      = capacity
        self.rate          = rate_per_second  # tokens per second
        self._tokens       = float(capacity)
        self._last_refill  = time.monotonic()
        self._lock         = asyncio.Lock()

    async def consume(self) -> bool:
        async with self._lock:
            now     = time.monotonic()
            elapsed = now - self._last_refill
            self._tokens = min(self.capacity, self._tokens + elapsed * self.rate)
            self._last_refill = now

            if self._tokens >= 1.0:
                self._tokens -= 1.0
                return True
            return False


# ---------------------------------------------------------------------------
# RetryPolicy
# ---------------------------------------------------------------------------

@dataclass
class RetryPolicy:
    max_attempts:       int   = 3
    initial_backoff:    float = 0.1   # seconds
    max_backoff:        float = 2.0   # seconds
    backoff_multiplier: float = 2.0
    hedge_delay:        Optional[float] = None  # seconds; None disables hedging
    hedge_bucket:       Optional[TokenBucket]    = None
    breaker:            Optional[CircuitBreaker] = None

    @classmethod
    def with_defaults(cls) -> "RetryPolicy":
        """Production-safe policy: 3 retries, 10 hedges/s, 5-failure circuit breaker."""
        return cls(
            breaker      = CircuitBreaker(max_failures=5, open_timeout=30.0, half_open_probes=2),
            hedge_bucket = TokenBucket(capacity=5, rate_per_second=10.0),
        )


# ---------------------------------------------------------------------------
# execute_with_retry
# ---------------------------------------------------------------------------

async def execute_with_retry(
    policy: RetryPolicy,
    operation: Callable[[], Awaitable[T]],
) -> T:
    """Run `operation` with the given RetryPolicy (backoff + hedging + circuit breaker)."""

    if policy.breaker and not await policy.breaker.allow():
        raise RuntimeError("circuit open: too many recent failures")

    if policy.hedge_delay is not None:
        return await _execute_with_hedging(policy, operation)

    return await _execute_retries_only(policy, operation)


async def _execute_retries_only(policy: RetryPolicy, operation: Callable[[], Awaitable[T]]) -> T:
    backoff   = policy.initial_backoff
    last_err: Optional[Exception] = None

    for attempt in range(1, policy.max_attempts + 1):
        try:
            result = await operation()
            if policy.breaker:
                await policy.breaker.record_success()
            return result
        except Exception as e:
            if policy.breaker:
                await policy.breaker.record_failure()
            last_err = e
            if attempt == policy.max_attempts:
                break
            # Full-jitter backoff: sleep rand[0, backoff)
            sleep_duration = random.uniform(0, backoff)
            await asyncio.sleep(sleep_duration)
            backoff = min(backoff * policy.backoff_multiplier, policy.max_backoff)

    raise last_err  # type: ignore[misc]


async def _execute_with_hedging(policy: RetryPolicy, operation: Callable[[], Awaitable[T]]) -> T:
    result_queue: asyncio.Queue = asyncio.Queue(maxsize=2)

    async def run_attempt():
        try:
            r = await _execute_retries_only(policy, operation)
            await result_queue.put(("ok", r))
        except Exception as e:
            await result_queue.put(("err", e))

    # Primary attempt
    asyncio.ensure_future(run_attempt())

    try:
        # Wait for primary to finish or hedge_delay to expire
        first = await asyncio.wait_for(result_queue.get(), timeout=policy.hedge_delay)
        status, value = first
        if status == "ok":
            return value  # type: ignore[return-value]
    except asyncio.TimeoutError:
        pass  # Hedge delay expired; fire duplicate

    # Rate-limit the hedge
    can_hedge = True
    if policy.hedge_bucket:
        can_hedge = await policy.hedge_bucket.consume()

    if can_hedge:
        asyncio.ensure_future(run_attempt())

    # Return whichever completes first with success
    for _ in range(2):
        status, value = await result_queue.get()
        if status == "ok":
            return value  # type: ignore[return-value]

    raise value  # type: ignore[misc]
